sklearn/cluster/_hdbscan/_boruvka.pyx

```diff
@@ -0,0 +1,760 @@
+) noexcept nogil:
+
+    cdef float64_t d, d1, d2, rdist = 0.0
+    cdef intp_t j
+
+    if metric.p == INF:
+        for j in range(num_features):
+            d1 = node_bounds[0, node1, j] - node_bounds[1, node2, j]
+            d2 = node_bounds[0, node2, j] - node_bounds[1, node1, j]
+            d = (d1 + fabs(d1)) + (d2 + fabs(d2))
+
+            rdist = max(rdist, 0.5 * d)
+    else:
+        # here we'll use the fact that x + abs(x) = 2 * max(x, 0)
+        for j in range(num_features):
+            d1 = node_bounds[0, node1, j] - node_bounds[1, node2, j]
+            d2 = node_bounds[0, node2, j] - node_bounds[1, node1, j]
+            d = (d1 + fabs(d1)) + (d2 + fabs(d2))
+
+            rdist += pow(0.5 * d, metric.p)
+
+    return metric._rdist_to_dist(rdist)
+
+
+cdef class BoruvkaUnionFind:
+    """
+    A union find implementation which avoids virtual nodes in order to keep track
+    of exact correspondence between initial elements and components.
+
+    Parameters
+    ----------
+
+    size : int
+        The total size of the set of objects to
+        track via the union find structure.
+
+    Attributes
+    ----------
+
+    is_component : array of bool; shape (size, 1)
+        Array specifying whether each element of the
+        set is the root node, or identifier for
+        a component.
+    """
+
+    cdef intp_t[::1] _parent
+    cdef uint8_t[::1] _rank
+    cdef uint8_t[::1] is_component
+
+    def __init__(self, size):
+        self._parent = np.arange(size, dtype=np.intp)
+        self._rank = np.zeros(size, dtype=np.uint8)
+        self.is_component = np.ones(size, dtype=np.uint8)
+
+    cdef int union_(self, intp_t x, intp_t y) noexcept nogil:
+        """Union together elements x and y"""
+        cdef intp_t x_root = self.find(x)
+        cdef intp_t y_root = self.find(y)
+
+        if x_root == y_root:
+            return 0
+
+        if self._rank[x_root] < self._rank[y_root]:
+            self._parent[x_root] = y_root
+            self.is_component[x_root] = 0
+        elif self._rank[x_root] > self._rank[y_root]:
+            self._parent[y_root] = x_root
+            self.is_component[y_root] = 0
+        else:
+            self._rank[x_root] += 1
+            self._parent[y_root] = x_root
+            self.is_component[y_root] = 0
+
+        return 0
+
+    cdef intp_t find(self, intp_t x) noexcept nogil:
+        """Find the root or identifier for the component that x is in"""
+        cdef intp_t x_parent
+        cdef intp_t x_grandparent
+
+        x_parent = self._parent[x]
+        while True:
+            if x_parent == x:
+                return x
+            x_grandparent = self._parent[x_parent]
+            self._parent[x] = x_grandparent
+            x = x_parent
+            x_parent = x_grandparent
+
+    cdef cnp.ndarray[intp_t, ndim=1] components(self):
+        """Return an array of all component roots/identifiers"""
+        return np.array(self.is_component).nonzero()[0]
```

Comment:
I am wondering whether we could make this lazy/online so that we do not need to allocate a new array in a `gil`-context every-time it is called.

For instance, we could:
 - allocate a a single array which could be partitioned in two halves with $\mathcal{O}(1)$ swaps with a first half for components being returned as a view. This would be efficient due to data locality, would preserve the logic of callers operating on arrays, but would not respect ordering.
 - create two link-lists, one for identifying components and the other with other nodes we could update in $O(1)$. This would not be as efficient since it break data locality, would need adapting the logic of callers so that it operates on a link-list, but would respect ordering.

What do you think?